### DISCUSSION TITLE
Understanding "temperature" parameter for WFST decoding

### DISCUSSION THREAD

**Discussion by Jesteinbe** [NVIDIA-NeMo/NeMo#7787] [Q&A]
Bottom line up front: I'm trying to understand the differences between the various ways you can specify "temperature" in a Fastconformer hybrid RNNT-CTC model and what they mean.

I've been working with some coworkers to use riva's wfst decoder (https://github.com/nvidia-riva/riva-asrlib-decoder/blob/main/src/riva/asrlib/decoder/test_decoder.py). I trained some models that were fine tuned from https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/stt_multilingual_fastconformer_hybrid_large_pc but i've been getting an error when i try to use the WFST decoder:

```
File "/home/.conda/envs/riva/lib/python3.10/site-packages/nemo/collections/asr/parts/utils/asr_confidence_utils.py", line 172, in _init_confidence_measure
    self.temperature = confidence_method_cfg.temperature
AttributeError: 'str' object has no attribute 'temperature'
```

I found a model that did work and discovered the primary difference between our model configs is that mine says 
```
decoding:
  confidence_cfg:
    method_cfg: DEPRECATED
```

whereas the model that does work has:
```
decoding:
  confidence_cfg:
    method_cfg:
      name: entropy
      entropy_type: tsallis
      temperature: 0.33
      entropy_norm: exp
```

As far as i can see, temperature can be specified in multiple places in the config though many are set to `DEPRECATED`, e.g. decoding.confidence_cfg.measure_cfg.temperature=DEPRECATED, decoding.greedy.confidence_measure_cfg.temperature=DEPRECATED, decoding.beam.softmax_temperature=1.0, decoding.temperature=1.0 (and the same flags exist under aux_ctc as well).

So my question is this: what do the various `temperature` fields actually refer to and which one(s) are in active use? Any advice will be helpful. Thanks!

**Comment by titu1994** [comment]
@Kipok @GNroy

**Comment by GNroy** [comment]
We decided to separate `temperature` from `method_cfg` from the ensemble-related temperature.
You can use `alpha` instead of `temperature` in `method_cfg`, see https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/utils/asr_confidence_utils.py
